overlord/devicestate/helpers.go

```diff
@@ -41,3 +49,53 @@ func gadgetDataFromInfo(info *snap.Info, constraints *gadget.ModelConstraints) (
 	}
 	return &gadget.GadgetData{Info: gi, RootDir: info.MountDir()}, nil
 }
+
+// partitionFromLabel returns the node of the block device used by the filesystem
+// with the specified label.
+func partitionFromLabel(label string) (string, error) {
+	output, err := exec.Command("findfs", "LABEL="+label).CombinedOutput()
+	if err != nil {
+		return "", osutil.OutputErr(output, err)
+	}
+	dev := strings.TrimSpace(string(output))
+	return dev, nil
+}
+
+var (
+	sysClassBlock = "/sys/class/block"
+	devBlock      = "/dev/block"
+)
+
+// diskFromPartition returns the node of the disk device that contains the
+// specified partition.
+func diskFromPartition(part string) (string, error) {
```

Comment:
The gadget package has some code for poking around sys in `findParentDeviceWithWritableFallback()`. 
Perhaps it'd make more sense to move `diskFromPartition()` to the gadget package to keep it in one place.

The code finds the immediate parent device. Do we need it to play nice with devices that are behind a device mapper?

Reply:
In the specific case of this installer `ubuntu-seed` won't be behind a device mapper, but the code would indeed fail in such case. Finding the parent of a device-mapped device is tricky, it could reside in different physical disks so I think we could add a warning to the source code and assume that in our case it won't happen.

Reply:
If moving `diskFromPartition()` to the gadget package, perhaps we could move `partitionFromLabel()` as well since it's already implemented using mostly gadget calls? Or we can do this refactoring in a separate PR after the install handler lands.

Reply:
Actually we can't, in this situation, get the partition from the `ubuntu-seed` label because the `system-seed` partition shouldn't set a label (see commit 4c5a374). Get the disk device using the structure role instead.